ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java

```diff
@@ -96,6 +99,17 @@ protected AbstractBlockProcessor(
     this.protocolSchedule = protocolSchedule;
   }
 
+  private BlockAwareOperationTracer getBlockImportTracer(
+      final ServiceManager pluginContext, final BlockHeader header) {
+
+    // if we have a BlockImportTracerProvider from pluginContext, use it.
+    return Optional.ofNullable(pluginContext)
+        .flatMap(serviceManager -> serviceManager.getService(BlockImportTracerProvider.class))
+        .map(z -> z.getBlockImportTracer(header))
+        // otherwise return NO_TRACING
+        .orElse(BlockAwareOperationTracer.NO_TRACING);
```

Comment:
not blocking: can the resolution of the tracer to use be done only once since I assume the tracer will not change?

Reply:
I see, if it is not straight forward it could stay like that or just use a memoizing supplier